File: tools/cotizar/Generadores/registrar-cotizacion.py
#!/usr/bin/env python3
"""
Registra en el CRM (Twenty, crm.carbonbox.app) que se envió una cotización.

Busca o crea la Empresa + Oportunidad del cliente y la mueve a la etapa
"Propuesta enviada" (PROPUESTA_ENVIADA), con una nota con plan, precio y NIT.
Replica los patrones de crm-scripts/crm_lib.py y lead_intake.py del repo
carbonbox-crm (solo como referencia — este script vive en tools/cotizar,
NO modifica los scripts de infraestructura del CRM).

⚠️ REQUIERE UN TOKEN LOCAL — nunca lo pegues en el chat con Claude.
1. Genera un token de API en Twenty: crm.carbonbox.app → Configuración →
   API & Webhooks → API keys.
2. Guárdalo en un archivo `.env` en esta misma carpeta (Generadores/), así:
     TWENTY_API_TOKEN=el_token_aqui
3. Este archivo `.env` es local — no lo subas al repo (agrégalo a .gitignore
   si no está ya).

Uso:
    python3 registrar-cotizacion.py --cliente "Hotel Waya Guajira" \\
        --nit "900.123.456-7" --plan Pro --precio 4101 \\
        --servicio "Estimación de huella de carbono organizacional" \\
        --nota "Cotización enviada por correo el 16 de julio de 2026"

Nota: el mapeo de campos/mutaciones GraphQL sigue el patrón observado en
crm_lib.py y lead_intake.py del repo, pero no se ha probado contra la
instancia real de Twenty (requiere el token). Primera corrida: verifica
que la oportunidad y la nota queden bien en crm.carbonbox.app antes de
usarlo como parte del flujo habitual.
"""

import argparse
import json
import os
import sys
import urllib.request
from datetime import datetime, timezone

CORE = "https://crm.carbonbox.app/graphql"
HERE = os.path.dirname(os.path.abspath(__file__))
ENV_FILE = os.path.join(HERE, ".env")
TOKEN_FILE_ALT = os.path.join(HERE, "token crm.txt")  # formato simple: solo el token, sin prefijo


def cargar_token():
    tok = os.environ.get("TWENTY_API_TOKEN")
    if tok:
        return tok.strip()
    if os.path.exists(ENV_FILE):
        with open(ENV_FILE) as f:
            for line in f:
                if line.strip().startswith("TWENTY_API_TOKEN="):
                    return line.strip().split("=", 1)[1].strip()
    if os.path.exists(TOKEN_FILE_ALT):
        with open(TOKEN_FILE_ALT) as f:
            tok = f.read().strip()
        if tok:
            return tok
    sys.exit(
        "❌ No encontré el token de Twenty.\n"
        f"   Crea el archivo {ENV_FILE} con la línea:\n"
        "   TWENTY_API_TOKEN=tu_token_aqui\n"
        f"   (o un archivo {TOKEN_FILE_ALT} con solo el token)\n"
        "   Genera el token en crm.carbonbox.app → Configuración → API & Webhooks"
    )


def gql(query, variables=None):
    token = cargar_token()
    body = json.dumps({"query": query, "variables": variables or {}}).encode()
    req = urllib.request.Request(CORE, data=body, headers={
        "Authorization": f"Bearer {token}", "Content-Type": "application/json"})
    with urllib.request.urlopen(req, timeout=30) as r:
        out = json.load(r)
    if out.get("errors"):
        raise RuntimeError(json.dumps(out["errors"], ensure_ascii=False)[:800])
    return out["data"]


def find_or_create_company(nombre):
    d = gql("""query($n: String!) { companies(filter:{name:{ilike:$n}}, first:1) {
        edges { node { id name } } } }""", {"n": nombre})
    if d["companies"]["edges"]:
        return d["companies"]["edges"][0]["node"]["id"]
    d = gql("""mutation($data: CompanyCreateInput!) { createCompany(data:$data) { id } }""",
            {"data": {"name": nombre}})
    return d["createCompany"]["id"]


def find_open_opportunity(company_id):
    """Busca una oportunidad abierta (no ganada/perdida) de esta empresa."""
    d = gql("""query($cid: UUID!) { opportunities(
        filter: { companyId: { eq: $cid },
                  stage: { in: ["LEAD_CAPTURADO","CALIFICACION_BANT","DEMO",
                                 "PILOTO_45D","PROPUESTA_ENVIADA","EN_NEGOCIACION",
                                 "RENOVACION","NURTURING"] } }, first: 1) {
        edges { node { id name stage } } } }""", {"cid": company_id})
    edges = d["opportunities"]["edges"]
    return edges[0]["node"] if edges else None


def crear_oportunidad(company_id, nombre_cliente, servicio, precio):
    odata = {"name": f"{nombre_cliente} — {servicio}", "stage": "PROPUESTA_ENVIADA",
              "companyId": company_id}
    if precio:
        odata["amount"] = {"amountMicros": int(round(precio * 1_000_000)), "currencyCode": "USD"}
    d = gql("""mutation($data: OpportunityCreateInput!) { createOpportunity(data:$data) { id } }""",
            {"data": odata})
    return d["createOpportunity"]["id"]


def mover_a_propuesta_enviada(opp_id, precio):
    data = {"stage": "PROPUESTA_ENVIADA"}
    if precio:
        data["amount"] = {"amountMicros": int(round(precio * 1_000_000)), "currencyCode": "USD"}
    gql("""mutation($id: UUID!, $data: OpportunityUpdateInput!) {
        updateOpportunity(id:$id, data:$data) { id } }""", {"id": opp_id, "data": data})


def agregar_nota(opp_id, texto):
    d = gql("""mutation($data: NoteCreateInput!) { createNote(data:$data) { id } }""",
            {"data": {"title": "Cotización enviada", "bodyV2": {"markdown": texto}}})
    note_id = d["createNote"]["id"]
    gql("""mutation($data: NoteTargetCreateInput!) { createNoteTarget(data:$data) { id } }""",
        {"data": {"noteId": note_id, "targetOpportunityId": opp_id}})


def registrar_cotizacion(cliente, nit, plan, precio, servicio, nota_extra=None):
    company_id = find_or_create_company(cliente)
    existente = find_open_opportunity(company_id)

    if existente:
        opp_id = existente["id"]
        mover_a_propuesta_enviada(opp_id, precio)
        accion = f"Oportunidad existente actualizada → Propuesta enviada ({existente['name']})"
    else:
        opp_id = crear_oportunidad(company_id, cliente, servicio, precio)
        accion = "Oportunidad nueva creada en Propuesta enviada"

    fecha = datetime.now(timezone.utc).astimezone().strftime("%d de %B de %Y")
    texto = (f"**Servicio:** {servicio}\n"
             f"**Plan:** {plan}\n"
             f"**NIT:** {nit}\n"
             + (f"**Precio:** ${precio:,.0f} USD\n" if precio else "")
             + f"**Fecha de envío:** {fecha}\n"
             + (f"\n{nota_extra}" if nota_extra else ""))
    agregar_nota(opp_id, texto)

    print(f"✅ {accion}")
    print(f"   Oportunidad: {opp_id}")
    print(f"   CRM: https://crm.carbonbox.app/object/opportunity/{opp_id}")


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Registra una cotización enviada en el CRM CarbonBox")
    parser.add_argument("--cliente", required=True, help="Nombre de la empresa cliente")
    parser.add_argument("--nit", required=True, help="NIT del cliente")
    parser.add_argument("--plan", required=True, help="Esencial / Pro / Experto")
    parser.add_argument("--precio", type=float, help="Precio final en USD")
    parser.add_argument("--servicio", default="Estimación de huella de carbono",
                        help="Descripción del servicio cotizado")
    parser.add_argument("--nota", help="Texto adicional para la nota del CRM")

    args = parser.parse_args()
    registrar_cotizacion(args.cliente, args.nit, args.plan, args.precio, args.servicio, args.nota)

File: tools/cotizar/Generadores/registrar-cotizacion.js
#!/usr/bin/env node
/**
 * Registra en el CRM (Twenty, crm.carbonbox.app) que se envió una cotización.
 *
 * Busca o crea la Empresa + Oportunidad del cliente y la mueve a la etapa
 * "Propuesta enviada" (PROPUESTA_ENVIADA), con una nota con plan, precio y NIT.
 * Replica los patrones de crm-scripts/crm_lib.py y lead_intake.py del repo
 * carbonbox-crm (solo como referencia — este script vive en tools/cotizar,
 * NO modifica los scripts de infraestructura del CRM).
 *
 * ⚠️ REQUIERE UN TOKEN LOCAL — nunca lo pegues en el chat con Claude.
 * Ya debe existir en esta carpeta como "token crm.txt" (o un archivo .env
 * con la línea TWENTY_API_TOKEN=...). Ese archivo es local, no lo subas al repo.
 *
 * Uso (desde esta carpeta, en PowerShell o cmd):
 *   node registrar-cotizacion.js --cliente "Hotel Waya Guajira" --nit "900.123.456-7" --plan Pro --precio 4101 --servicio "Estimación de huella de carbono organizacional" --nota "Cotización enviada por correo el 16 de julio de 2026"
 *
 * Nota: el mapeo de campos/mutaciones GraphQL sigue el patrón observado en
 * crm_lib.py y lead_intake.py del repo. Primera corrida: verifica que la
 * oportunidad y la nota queden bien en crm.carbonbox.app.
 */

const fs = require("fs");
const path = require("path");

const CORE = "https://crm.carbonbox.app/graphql";
const HERE = __dirname;
const ENV_FILE = path.join(HERE, ".env");
const TOKEN_FILE_ALT = path.join(HERE, "token crm.txt"); // formato simple: solo el token

function cargarToken() {
  if (process.env.TWENTY_API_TOKEN) return process.env.TWENTY_API_TOKEN.trim();
  if (fs.existsSync(ENV_FILE)) {
    const txt = fs.readFileSync(ENV_FILE, "utf8");
    const m = txt.split(/\r?\n/).find(l => l.trim().startsWith("TWENTY_API_TOKEN="));
    if (m) return m.trim().split("=").slice(1).join("=").trim();
  }
  if (fs.existsSync(TOKEN_FILE_ALT)) {
    const tok = fs.readFileSync(TOKEN_FILE_ALT, "utf8").trim();
    if (tok) return tok;
  }
  console.error("❌ No encontré el token de Twenty.");
  console.error(`   Crea el archivo ${ENV_FILE} con la línea: TWENTY_API_TOKEN=tu_token_aqui`);
  console.error(`   (o un archivo ${TOKEN_FILE_ALT} con solo el token)`);
  process.exit(1);
}

async function gql(query, variables = {}) {
  if (typeof fetch === "undefined") {
    console.error("❌ Tu versión de Node no tiene fetch incorporado (necesitas Node 18+).");
    console.error("   Dime qué versión tienes (node --version) y lo adapto.");
    process.exit(1);
  }
  const token = cargarToken();
  const res = await fetch(CORE, {
    method: "POST",
    headers: { "Authorization": `Bearer ${token}`, "Content-Type": "application/json" },
    body: JSON.stringify({ query, variables })
  });
  const out = await res.json();
  if (out.errors) {
    throw new Error(JSON.stringify(out.errors).slice(0, 800));
  }
  return out.data;
}

// El NIT se guarda en Company.nit, que es de donde lo lee el flujo /cotizar al preparar
// la siguiente cotización. Antes solo quedaba escrito dentro del texto de la nota, así que
// una empresa creada por este script quedaba sin NIT utilizable.
async function findOrCreateCompany(nombre, nit) {
  let d = await gql(`query($n: String!) { companies(filter:{name:{ilike:$n}}, first:1) {
      edges { node { id name nit } } } }`, { n: nombre });
  if (d.companies.edges.length) {
    const c = d.companies.edges[0].node;
    // Si ya tiene NIT no lo pisamos: el del CRM manda sobre el que venga por argumento.
    if (nit && !c.nit) {
      await gql(`mutation($id: UUID!, $data: CompanyUpdateInput!) {
          updateCompany(id:$id, data:$data) { id } }`, { id: c.id, data: { nit } });
    }
    return c.id;
  }
  const data = { name: nombre };
  if (nit) data.nit = nit;
  d = await gql(`mutation($data: CompanyCreateInput!) { createCompany(data:$data) { id } }`, { data });
  return d.createCompany.id;
}

async function findOpenOpportunity(companyId) {
  const d = await gql(`query($cid: UUID!) { opportunities(
      filter: { companyId: { eq: $cid },
                stage: { in: ["LEAD_CAPTURADO","CALIFICACION_BANT","DEMO",
                               "PILOTO_45D","PROPUESTA_ENVIADA","EN_NEGOCIACION",
                               "RENOVACION","NURTURING"] } }, first: 1) {
      edges { node { id name stage } } } }`, { cid: companyId });
  const edges = d.opportunities.edges;
  return edges.length ? edges[0].node : null;
}

async function crearOportunidad(companyId, nombreCliente, servicio, precio) {
  const odata = { name: `${nombreCliente} — ${servicio}`, stage: "PROPUESTA_ENVIADA", companyId };
  if (precio) odata.amount = { amountMicros: Math.round(precio * 1_000_000), currencyCode: "USD" };
  const d = await gql(`mutation($data: OpportunityCreateInput!) { createOpportunity(data:$data) { id } }`,
    { data: odata });
  return d.createOpportunity.id;
}

// La calculadora maneja Esencial/Pro/Experto; el campo planCarbonbox del CRM tiene
// ESENCIAL/PRO/PREMIUM/ENTERPRISE/LICITACION. "Experto" se guarda como PREMIUM.
const PLAN_A_CODIGO = { esencial: "ESENCIAL", pro: "PRO", experto: "PREMIUM" };

function planACodigo(plan) {
  if (!plan) return null;
  return PLAN_A_CODIGO[String(plan).trim().toLowerCase()] || null;
}

async function moverAPropuestaEnviada(oppId, precio, extras = {}) {
  const data = { stage: "PROPUESTA_ENVIADA" };
  if (precio) data.amount = { amountMicros: Math.round(precio * 1_000_000), currencyCode: "USD" };
  const codigo = planACodigo(extras.plan);
  if (codigo) data.planCarbonbox = codigo;
  if (extras.linkCotizacion) {
    data.linkCotizacion = { primaryLinkUrl: extras.linkCotizacion, primaryLinkLabel: "Cotización" };
  }
  if (extras.borrador) data.borradorCorreo = extras.borrador;
  await gql(`mutation($id: UUID!, $data: OpportunityUpdateInput!) {
      updateOpportunity(id:$id, data:$data) { id } }`, { id: oppId, data });
}

async function agregarNota(oppId, texto) {
  const d = await gql(`mutation($data: NoteCreateInput!) { createNote(data:$data) { id } }`,
    { data: { title: "Cotización enviada", bodyV2: { markdown: texto } } });
  const noteId = d.createNote.id;
  await gql(`mutation($data: NoteTargetCreateInput!) { createNoteTarget(data:$data) { id } }`,
    { data: { noteId, targetOpportunityId: oppId } });
}

async function registrarCotizacion({ cliente, nit, plan, precio, servicio, nota, linkCotizacion, borrador }) {
  const companyId = await findOrCreateCompany(cliente, nit);
  const existente = await findOpenOpportunity(companyId);
  const extras = { plan, linkCotizacion, borrador };

  let oppId, accion;
  if (existente) {
    oppId = existente.id;
    await moverAPropuestaEnviada(oppId, precio, extras);
    accion = `Oportunidad existente actualizada → Propuesta enviada (${existente.name})`;
  } else {
    oppId = await crearOportunidad(companyId, cliente, servicio, precio);
    await moverAPropuestaEnviada(oppId, precio, extras);
    accion = "Oportunidad nueva creada en Propuesta enviada";
  }

  const fecha = new Date().toLocaleDateString("es-CO", { day: "numeric", month: "long", year: "numeric" });
  let texto = `**Servicio:** ${servicio}\n**Plan:** ${plan}\n**NIT:** ${nit}\n`;
  if (precio) texto += `**Precio:** $${Number(precio).toLocaleString("en-US")} USD\n`;
  texto += `**Fecha de envío:** ${fecha}\n`;
  if (nota) texto += `\n${nota}`;
  await agregarNota(oppId, texto);
  return { oppId, accion, texto };
}

function parseArgs(argv) {
  const out = {};
  for (let i = 0; i < argv.length; i++) {
    if (argv[i].startsWith("--")) {
      const key = argv[i].slice(2);
      const val = argv[i + 1];
      out[key] = val;
      i++;
    }
  }
  return out;
}

// Solo corre como CLI. Al importarlo (los tests) no ejecuta nada contra el CRM.
if (require.main === module) {
  const args = parseArgs(process.argv.slice(2));
  if (!args.cliente || !args.nit || !args.plan) {
    console.error("Uso: node registrar-cotizacion.js --cliente \"Nombre\" --nit \"900...\" --plan Pro [--precio 4101] [--servicio \"...\"] [--nota \"...\"]");
    process.exit(1);
  }

  // El borrador se pasa por ARCHIVO, no por argumento: es largo y multilínea, y como
  // argumento de shell se rompe con las comillas y los saltos de línea.
  let borrador = null;
  if (args["borrador-archivo"]) borrador = fs.readFileSync(args["borrador-archivo"], "utf8");

  registrarCotizacion({
    cliente: args.cliente,
    nit: args.nit,
    plan: args.plan,
    precio: args.precio ? parseFloat(args.precio) : null,
    servicio: args.servicio || "Estimación de huella de carbono",
    nota: args.nota || null,
    linkCotizacion: args["link-cotizacion"] || null,
    borrador
  }).then(({ oppId, accion }) => {
    console.log(`✅ ${accion}`);
    console.log(`   Oportunidad: ${oppId}`);
    console.log(`   CRM: https://crm.carbonbox.app/object/opportunity/${oppId}`);
  }).catch(e => { console.error("ERROR:", e.message); process.exit(1); });
}

module.exports = {
  registrarCotizacion, findOrCreateCompany, findOpenOpportunity,
  crearOportunidad, moverAPropuestaEnviada, agregarNota, parseArgs, planACodigo,
};
